projects/portal/src/app/pages/faucet/faucet.usecase.service.ts

```diff
@@ -1,51 +1,35 @@
 import { Injectable } from '@angular/core';
 import { ActivatedRoute } from '@angular/router';
-import { Config, ConfigService } from 'projects/portal/src/app/models/config.service';
-import { BehaviorSubject, combineLatest, Observable } from 'rxjs';
-import { first, map } from 'rxjs/operators';
+import { Config, ConfigService } from '../../models/config.service';
+import { combineLatest, Observable } from 'rxjs';
+import { map } from 'rxjs/operators';
 
 @Injectable({ providedIn: 'root' })
 export class FaucetUseCaseService {
-
-  //private config$: Observable<Config | undefined>;
-
   constructor(
     private readonly route: ActivatedRoute,
     private configService: ConfigService,
-  ) {
-    //this.config$ = this.configService.config$;
-    //this.denom$ = new BehaviorSubject('');
-  }
+  ) { }
 
   get config$(): Observable<Config | undefined> {
     return this.configService.config$;
   }
-
-
   get denoms$(): Observable<string[] | undefined> {
     return this.config$.pipe(
       map((config) => config?.extension?.faucet?.map((faucet) => faucet.denom)),
     );
   }
-
-  get address$(): Observable<string> {
-    return this.route.queryParams.pipe(map((queryParams) => queryParams.address));
+  get address$(): Observable<string | undefined> {
+    return this.route.queryParams.pipe(map((queryParams) => queryParams?.address ? queryParams.address : undefined));
   }
-
-  get amount$(): Observable<number> {
-    return this.route.queryParams.pipe(map((queryParams) => queryParams.amount));
+  get amount$(): Observable<number | undefined> {
+    return this.route.queryParams.pipe(map((queryParams) => queryParams?.amount ? queryParams.amount : undefined));
   }
-
-  get denom$(): BehaviorSubject<string> {
-    this.route.queryParams
-      .pipe(first())
-      .toPromise()
-      .then((params) => this.denom$?.next(params.denom));
-    return this.denom$
+  get denom$(): Observable<string> {
+    return this.route.queryParams.pipe(map((queryParams) => queryParams?.denom ? queryParams.denom : undefined));
   }
-
   get creditAmount$(): Observable<number> {
```

Comment:
For example, now `creditAmount$` depends on `denom$` and `denom$` depends on `ActivatedRoute`.
So if you give `denom$` as an argument, `creditAmount$` will no longer depend on `ActivatedRoute`.

```ts
creditAmount$(denom$: Observable<string[] | undefined> /*👈*/): Observable<number>{
	return combineLatest([this.config$, denom$ /*👈*/]).pipe(
      map(([config, denom]) => {
        const faucet = config?.extension?.faucet
          ? config.extension.faucet.find((faucet) => faucet.denom === denom)
          : undefined;
        const creditAmount = faucet ? faucet.creditAmount : 0;
        return creditAmount;
      }),
    );
}
```

(It is worth discussing whether the argument should be "Observable" or a simple value, but I think Observable is better here because it requires fewer changes.)